### PR TITLE
Registration - adding accounts to CredentialsManager Assignment - 22.11.2024

### PR DESCRIPTION
**Pull request by bakhodirsolieff**
**Changes made :**

- Successful registration(new user is present in CredentialsManager)
- Case Insensitive in email
- Email already used error

**Correct input values:**

- Password(digits)
- email format

"After successful registration, go to Login Activity"

### CODE REVIEW

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt
Comment (KamronGIT0319) [inline app/src/main/java/com/example/myapplication/CredentialsManager.kt:1]: I likel the idea of isemailalreadyused function which was not here before, and changing from class to object

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt
Comment (KamronGIT0319) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt:1]: Shortening the line 15 is really time saving  a

**Review by KamronGIT0319 — APPROVED** [review]

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt [hunk 1]
```diff
@@ -9,10 +10,20 @@ class CredentialsManager {
...
+
     fun isEmailValid(email: String): Boolean {
         return email.matches(emailPattern)
     }
 
+    fun isEmailAlreadyUsed(email: String): Boolean {
+        return registeredEmails.contains(email.lowercase())
+    }
+
+    fun registerUser(email: String) {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CredentialsManager.kt]: If we are to use that credentials in login, don't we need a password here too?

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt:1]: What about testing registerUser method?

**Review by mtkar-dev — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt
Comment (bakhodirsolieff) [reply app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt:1]: fixed that by adding registerUser method in test class

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 2]
```diff
@@ -68,53 +68,53 @@ class SignUpDetailsActivity : AppCompatActivity() {
...
         clearError(passwordLayout)
 
-        if (fullName.isEmpty()) {
-            setError(fullNameLayout, getString(R.string.error_name_required))
-            return
-        }
-
-        if (email.isEmpty() || !credentialsManager.isEmailValid(email)) {
-            setError(emailLayout, getString(R.string.error_invalid_email))
+        if (!credentialsManager.ValidateCredentialsForSignUp(
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt]: methods should be lowerCamelCase

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 3]
```diff
@@ -68,53 +68,53 @@ class SignUpDetailsActivity : AppCompatActivity() {
...
+                showToast(getString(R.string.error_terms_and_conditions_required))
+            }
             return
         }
 
-        if (phone.isEmpty() || !credentialsManager.isValidPhoneNumber(phone)) {
-            setError(phoneLayout, getString(R.string.error_phone_number_required))
-            return
-        }
+        proceedToNextStep(email, password)
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt]: Can you find a better name for that method?

File: app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt [hunk 4]
```diff
@@ -68,53 +68,53 @@ class SignUpDetailsActivity : AppCompatActivity() {
...
         }
 
-        if (!isTermsAccepted) {
-            showToast(getString(R.string.error_terms_and_conditions_required))
-            return
+        if (credentialsManager.registerUser(email, password)) {
+            showToast(getString(R.string.success_registration))
+            navigateToLogin()
+        } else {
+            setError(emailLayout, getString(R.string.error_email_already_used))
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/SignUpDetailsActivity.kt]: Duplicated code for displaying email already used
Can be simplified

**Review by mtkar-dev — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/myapplication/FragmentA.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package com.example.myapplication
+
+import androidx.fragment.app.Fragment
+
+
+public class FragmentA : Fragment(R.layout.fragment_a) {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/FragmentA.kt]: In Kotlin you don't need {} after constructor if the class body is empty

`public class FragmentA : Fragment(R.layout.fragment_a)`
is just enough

File: app/src/main/java/com/example/myapplication/FragmentA.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.example.myapplication
+
+import androidx.fragment.app.Fragment
+
+
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/FragmentA.kt]: Unnecessary empty line

File: app/src/main/java/com/example/myapplication/CredentialsManager.kt [hunk 7]
```diff
@@ -9,16 +10,34 @@ class CredentialsManager {
...
-        return isEmailValid(email) && isValidPassword(password) && isCheckboxChecked
+    fun registerUser(email: String, password: String): Boolean {
+        if (isEmailAlreadyUsed(email)) {
+            return false
+        }
+        registeredUsers[email.lowercase()] = password
+        return true
+    }
+
+    fun validateLogin(email: String, password: String): Boolean {
```
Comment (mtkar-dev) [inline app/src/main/java/com/example/myapplication/CredentialsManager.kt:35]: I don't think you validate login here - but just login instead
I know we don't set the flag if the user is logged in yet, but generally the goal is to log in the user, not to see if he can log in

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -19,7 +20,7 @@ androidx-appcompat = { group = "androidx.appcompat", name = "appcompat", version
 material = { group = "com.google.android.material", name = "material", version.ref = "material" }
 androidx-activity = { group = "androidx.activity", name = "activity", version.ref = "activity" }
 androidx-constraintlayout = { group = "androidx.constraintlayout", name = "constraintlayout", version.ref = "constraintlayout" }
-
+androidx-fragments = { group = "androidx.fragment", name = "fragment-ktx", version.ref = "fragments"}
```
Comment (mtkar-dev) [inline gradle/libs.versions.toml]: Missing empty line below

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt [hunk 9]
```diff
@@ -21,20 +21,19 @@ class CredentialsManagerTest {
...
         // Invalid email
-        assertFalse(credentialsManager.validateCredentials("invalid-email", "password123", true))
+        assertFalse(credentialsManager.validateCredentials("invalid-email", "password123"))
 
         // Invalid password
-        assertFalse(credentialsManager.validateCredentials("example@test.com", "short", true))
+        assertFalse(credentialsManager.validateCredentials("example@test.com", "short"))
 
         // Checkbox not checked
         assertFalse(
```
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt]: If you removed the checkboxChecked check - that seems obsolete
Either the comment is misleading or that assertion is obsolete

**Review by mtkar-dev — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml [hunk 10]
```diff
@@ -19,8 +20,7 @@ androidx-appcompat = { group = "androidx.appcompat", name = "appcompat", version
 material = { group = "com.google.android.material", name = "material", version.ref = "material" }
 androidx-activity = { group = "androidx.activity", name = "activity", version.ref = "activity" }
 androidx-constraintlayout = { group = "androidx.constraintlayout", name = "constraintlayout", version.ref = "constraintlayout" }
-
+androidx-fragments = { group = "androidx.fragment", name = "fragment-ktx", version.ref = "fragments" }
```
Comment (mtkar-dev) [inline gradle/libs.versions.toml]: You forgot to remove those from your PR

File: app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt [hunk 11]
```diff
@@ -21,22 +21,13 @@ class CredentialsManagerTest {
     @Test
     fun testValidateCredentials() {
         // Valid email, password, and checkbox checked
```
Comment (mtkar-dev) [inline app/src/test/java/com/example/myapplication/CredentialsManagerTest.kt]: My bad, didn't see it
Another remain of checkbox checked

**Review by mtkar-dev — APPROVED** [review]
As in other case - remove what I commented and you're good to merge

File: app/build.gradle.kts [hunk 12]
```diff
@@ -43,6 +43,7 @@ dependencies {
     implementation("com.google.android.material:material:1.4.0")
     implementation(libs.androidx.activity)
     implementation(libs.androidx.constraintlayout)
+    implementation(libs.androidx.fragments)
```
Comment (mtkar-dev) [inline app/build.gradle.kts]: Also here

Comment (bakhodirsolieff) [reply app/build.gradle.kts]: Thanks for the review, i deleted that

**Review by mtkar-dev — APPROVED** [review]

**Review by KamronGIT0319 — APPROVED** [review]

**Review by KamronGIT0319 — APPROVED** [review]
.

File: app/src/main/AndroidManifest.xml [hunk 13]
```diff
@@ -15,7 +15,7 @@
```
Comment (KamronGIT0319) [inline app/src/main/AndroidManifest.xml:15]: .

File: app/src/main/java/com/example/myapplication/CreateAccountActivity.kt [hunk 14]
```diff
@@ -3,15 +3,14 @@ package com.example.myapplication
 import android.content.Intent
```
Comment (KamronGIT0319) [inline app/src/main/java/com/example/myapplication/CreateAccountActivity.kt:3]: .

**Review by KamronGIT0319 — APPROVED** [review]

**Review by KamronGIT0319 — APPROVED** [review]
.

**Review by KamronGIT0319 — APPROVED** [review]
.

**Pull request merged**